core/cache/lxdprofilewatcher.go

```diff
@@ -174,35 +175,40 @@ func (w *MachineAppLXDProfileWatcher) unitChange(topic string, value interface{}
 }
 
 func (w *MachineAppLXDProfileWatcher) addUnit(unit *Unit) bool {
-	_, ok := w.applications[unit.details.Application]
+	unitName := unit.Name()
+	appName := unit.Application()
+
+	_, ok := w.applications[appName]
 	if !ok {
-		curl := unit.details.CharmURL
+		curl := unit.CharmURL()
```

Comment:
I appreciate the functions nonetheless. having object X access Y.private.attribute is also something to avoid.